packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.tsx

```diff
@@ -0,0 +1,398 @@
+   */
+  private _defaultOnDecrement = (value: string) => {
+    let newValue = Math.max(Number(value) - this.props.step, this.props.min);
+    return String(newValue);
+  }
+
+  /**
+   * Set the label direction with the gap on the correct side
+   */
+  @autobind
+  private _labelDirectionHelper(): any {
+    let style: any = {};
+
+    switch (this.props.labelPosition) {
+      case Position.start:
+        style = { float: 'left', marginRight: this.props.labelGapSpace + 'px' };
+        break;
+      case Position.end:
+        style = { float: 'right', marginLeft: this.props.labelGapSpace + 'px' };
+        break;
+      case Position.top:
+        style = { marginBottom: this.props.labelGapSpace + 'px' };
+        break;
+      case Position.bottom:
+        style = { marginTop: this.props.labelGapSpace + 'px' };
+    }
+
+    return style;
+  }
+
+  private _onChange() {
+    /**
+     * A noop input change handler.
+     * https://github.com/facebook/react/issues/7027.
+     * Using the native onInput handler fixes the issue but onChange
+     * still need to be wired to avoid React console errors
+     * TODO: Check if issue is resolved when React 16 is available.
+     */
+  }
+
+  /**
+   * This is used when validating text entry
+   * in the input (not when changed via the buttons)
+   * @param event - the event that fired
+   */
+  @autobind
+  private _validate(event: React.FocusEvent<HTMLInputElement>) {
+    const element: HTMLInputElement = event.target as HTMLInputElement;
+    const value: string = element.value;
+    if (this.state.value) {
+      const newValue = this._onValidate(value);
+      this._lastValidValue = newValue;
+      this.setState({ value: newValue });
+    }
+  }
+
+  /**
+   * The method is needed to ensure we are updating the actual input value.
+   * without this our value will never change (and validation will not have the correct number)
+   * @param event - the event that was fired
+   */
+  @autobind
+  private _onInputChange(event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>): void {
+    const element: HTMLInputElement = event.target as HTMLInputElement;
+    const value: string = element.value;
+
+    this.setState({
+      value: value,
+    });
+  }
+
+  /**
+   * Update the value with the given stepFunction
+   * @param shouldSpin - should we fire off another updateValue when we are done here? This should be true
+   * when spinning in response to a mouseDown
+   * @param stepFunction - function to use to step by
+   */
+  @autobind
+  private _updateValue(shouldSpin: boolean, stepFunction: (string) => string) {
+    const newValue = stepFunction(this.state.value);
+    this._lastValidValue = newValue;
+    this.setState({ value: newValue });
+
+    if (this._spinning !== shouldSpin) {
+      this._spinning = shouldSpin;
+    }
+
+    if (shouldSpin) {
+      this._currentStepFunctionHandle = window.setTimeout(() => { this._updateValue(shouldSpin, stepFunction); }, this._stepDelay);
```

Comment:
this is still using setTimeout directly when it should be using _async:


  /**
   * Gets the async instance associated with the component, created on demand. The async instance gives
   * subclasses a way to execute setTimeout/setInterval async calls safely, where the callbacks
   * will be cleared/ignored automatically after unmounting. The helpers within the async object also
   * preserve the this pointer so that you don't need to "bind" the callbacks.